packages/amplify-graphql-schema-generator/src/schema-generator/schema-overrides.ts

```diff
@@ -18,16 +20,54 @@ export const applySchemaOverrides = (document: DocumentNode, existingDocument?:
         const correspondingField = findMatchingField(node, parentObjectType, existingDocument);
         if (!correspondingField) return;
 
+        // eslint-disable-next-line consistent-return
         return applyFieldOverrides(node, correspondingField);
       },
     },
   };
 
-  const updatedDocument = visit(document, schemaVisitor);
+  updatedDocument = visit(updatedDocument, schemaVisitor);
   updatedDocument['definitions'] = [...updatedDocument['definitions'], ...getNonModelTypes(existingDocument)];
+
   return updatedDocument;
 };
 
+const preserveRelationalDirectives = (document: DocumentNode, existingDocument: DocumentNode): DocumentNode => {
+  const MODEL_DIRECTIVE_NAME = 'model';
+  const RELATIONAL_DIRECTIVES = ['hasOne', 'hasMany', 'belongsTo'];
+  const documentWrapper = document as any;
+
+  if (!existingDocument) {
+    return document;
+  }
+  existingDocument.definitions
+    .filter((def) => def.kind === 'ObjectTypeDefinition' && def.directives.find((dir) => dir.name.value === MODEL_DIRECTIVE_NAME))
+    .forEach((existingObject: ObjectTypeDefinitionNode) => {
+      const newObject = document.definitions.find(
+        (def) => def.kind === 'ObjectTypeDefinition' && def.name.value === existingObject.name.value,
+      ) as ObjectTypeDefinitionNode;
+      if (!newObject) {
+        return;
+      }
+      const relationalFields = existingObject.fields.filter((field) =>
+        field.directives.find((dir) => RELATIONAL_DIRECTIVES.includes(dir.name.value)),
+      );
+      const newObjectWrapper = new ObjectDefinitionWrapper(newObject);
+      relationalFields.forEach((relationalField: FieldDefinitionNode) => {
+        newObjectWrapper.fields.push(new FieldWrapper(relationalField));
+      });
+
+      if (relationalFields.length > 0) {
```

Comment:
This branch creates a potentially bizarre modal behavior. If I'm reading this right if I go make changes to my Model after import, then upon the next import those changes are scrubbed, but if those changes ALSO include a relational update then those changes are maintained (not just the relationship).

Is that the desired behavior, or should we be doing the merge one layer deeper (e.g. only maintain relational fields in the colliding models?

Reply:
That a great question. This logic copies only the relationship. `newObject` is the regenerated GraphQL type and `existingObject` is the current state of the GraphQL type from schema.rds.graphql file. We only copy the relational fields from existing type to the new type in this method (line 56-58).

I'm adding a test specific to this scenario.

However, I had other concerns that I will review with Rene and address in a follow up if required.
- What if the table belongs to the related type is deleted? Should we still retain the relationship?
- What if one of the references fields removed?

Reply:
Got it - thank you for explaining. I missed that on my first read, but makes sense after reviewing.